### ISSUE TITLE
Wayfire crash

### ISSUE THREAD

**Issue by eternalinflation** [WayfireWM/wayfire#1165]
**Wayfire crashes with `swaybg`, idle plugin and DPMS under certain circumstances**

**How to reproduce**
1. Start wayfire session via `wayfire-session` (env settings see below) from VT
2. Open `wcm`, in `Background View` plugin set `Command` to `swaybg -m fill -i` and `File` to `$HOME/.wallpaper`. Set `.wallpaper` as a symlink to an image with size/aspect ratio different than the output / screen (may or may not matter)
3. In Idle plugin set `DPMS timeout: 5` and  `DPMS screensaver: 10`
4. Let DPMS kick in and settle for 20 seconds or so
5. After Screen went into power save hit Enter key
6. Wayfire session crashes and returns to the VT

**Expected behavior**
Wayfire should not crash regardless of third party applications' behavior

**Wayfire version**
0.7.0 / 0.7.1 releases

**WCM version**
0.7.0 release

**stacktrace**
[wayfire-0.7.1/src/core/output-layout.cpp:467] disabling output: DP-1
[wayfire-0.7.1/src/core/output-layout.cpp:140] transfer views from DP-1 -> NOOP-1
[wayfire-0.7.1/src/main.cpp:141] Fatal error: Segmentation fault
#1  wf::print_trace(bool) ??:?
#2  killpg ??:?
#3  wayfire_background_view::fini() ??:?
#4  plugin_manager::destroy_plugin(std::unique_ptr<wf::plugin_interface_t, std::default_delete<wf::plugin_interface_t> >&) ??:?
#5  plugin_manager::deinit_plugins(bool) ??:?
#6  plugin_manager::~plugin_manager() ??:?
#7  wf::output_impl_t::~output_impl_t() ??:?
#8  wf::output_impl_t::~output_impl_t() ??:?
#9  wf::output_layout_output_t::destroy_wayfire_output() ??:?
#10 wf::output_layout_output_t::apply_state(wf::output_state_t const&) ??:?
#11 wf::output_layout_t::impl::apply_configuration(std::map<wlr_output*, wf::output_state_t, std::less<wlr_output*>, std::allocator<std::pair<wlr_output* const, wf::output_state_t> > > const&) ??:?
#12 wf::output_layout_t::impl::remove_output(wlr_output*) ??:?
#13 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > wf::log::to_string<char const*>(char const*) ??:?
#14 wlr_region_confine ??:?
#15 wlr_output_destroy ??:?
#16 wlr_drm_connector_get_id ??:?
#17 wlr_region_confine ??:?
#18 wlr_wl_input_device_get_seat ??:?
#19 wl_event_loop_dispatch ??:?
#20 wl_display_run ??:?
#21 main ??:?
#22 __libc_start_main ??:?
#23 _start ??:?

**wayfire-session**
#!/bin/bash
exec 1> >(logger --id=$$ --stderr --tag wayfire) 2>&1
[ -z "$XDG_DATA_DIRS" ] && export XDG_DATA_DIRS="/usr/local/share:/usr/share"
export XDG_SESSION_TYPE=wayland
export GDK_BACKEND=wayland
export QT_QPA_PLATFORM=wayland
export QT_QPA_PLATFORMTHEME=qt5ct
export MOZ_ENABLE_WAYLAND=1
export GDK_DPI_SCALE=1.5
export QT_AUTO_SCREEN_SCALE_FACTOR=0
export QT_FONT_DPI=144
exec /usr/bin/wayfire "$@"

**Comment by eternalinflation** [comment]
Here' another way to crash Wayfire:
1. Start wayfire via wayfire-session from VT
2. Switch monitor off - switch monitor on - switch monitor off - switch monitor on again
After 2nd cycle Wayfire crashes and returns to the VT. 

**Output from wayfire-session 2> wf.log**:
wayfire[1812835]: II 01-05-21 16:20:36.400 - [backend/drm/drm.c:1221] Scanning DRM connectors on /dev/dri/card0
wayfire[1812835]: II 01-05-21 16:20:36.400 - [backend/drm/drm.c:1403] 'DP-1' disconnected
wayfire[1812835]: II 01-05-21 16:20:36.400 - [wayfire-0.7.1/src/core/output-layout.cpp:1034] remove output: DP-1
wayfire[1812835]: II 01-05-21 16:20:36.400 - [wayfire-0.7.1/src/core/output-layout.cpp:980] new output: NOOP-1
wayfire[1812835]: II 01-05-21 16:20:36.400 - [wayfire-0.7.1/src/core/output-layout.cpp:388] loaded mode auto
wayfire[1812835]: II 01-05-21 16:20:36.400 - [wayfire-0.7.1/src/core/output-layout.cpp:577] Couldn't find matching mode 1200x720@60 for output NOOP-1. Trying t$
wayfire[1812835]: (type equals variant: [type: string, value: toplevel] | (type equals variant: [type: string, value: x-or] & focusable equals variant: [type: $
wayfire[1812835]: type equals variant: [type: string, value: overlay]
wayfire[1812835]: false
wayfire[1812835]: false
wayfire[1812835]: false
wayfire[1812835]: EE 01-05-21 16:20:36.404 - [wayfire-0.7.1/src/core/output-layout.cpp:467] disabling output: DP-1
wayfire[1812835]: II 01-05-21 16:20:36.404 - [wayfire-0.7.1/src/core/output-layout.cpp:140] transfer views from DP-1 -> NOOP-1
wayfire[1812835]: EE 01-05-21 16:20:36.405 - [wayfire-0.7.1/src/main.cpp:141] Fatal error: Segmentation fault
wayfire[1812835]: EE 01-05-21 16:20:36.413 - #1  wf::print_trace(bool) ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.421 - #2  killpg ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.428 - #3  wf::pointer_t::get_focus() const ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.435 - #4  wayfire_background_view::fini() ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.443 - #5  plugin_manager::destroy_plugin(std::unique_ptr<wf::plugin_interface_t, std::default_delete<wf::plugin_interfac$
wayfire[1812835]: EE 01-05-21 16:20:36.450 - #6  plugin_manager::deinit_plugins(bool) ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.459 - #7  plugin_manager::~plugin_manager() ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.467 - #8  wf::output_impl_t::~output_impl_t() ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.474 - #9  wf::output_impl_t::~output_impl_t() ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.481 - #10 wf::output_layout_output_t::destroy_wayfire_output() ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.488 - #11 wf::output_layout_output_t::apply_state(wf::output_state_t const&) ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.495 - #12 wf::output_layout_t::impl::apply_configuration(std::map<wlr_output*, wf::output_state_t, std::less<wlr_output*$
wayfire[1812835]: EE 01-05-21 16:20:36.502 - #13 wf::output_layout_t::impl::remove_output(wlr_output*) ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.509 - #14 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > wf::log::to_string<char const*$
wayfire[1812835]: EE 01-05-21 16:20:36.515 - #15 wlr_region_confine ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.521 - #16 wlr_output_destroy ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.527 - #17 wlr_drm_connector_get_id ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.533 - #18 wlr_region_confine ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.539 - #19 wlr_wl_input_device_get_seat ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.545 - #20 wl_event_loop_dispatch ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.551 - #21 wl_display_run ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.558 - #22 main ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.564 - #23 __libc_start_main ??:?
wayfire[1812835]: EE 01-05-21 16:20:36.571 - #24 _start ??:?

**Comment by unrelentingtech** [comment]
(Please use \`\`\` fenced code blocks for the stacktrace! Don't reference all the `#0..42` github issues :D)

This should probably be reported against `background-view` in https://github.com/WayfireWM/wayfire-plugins-extra. From a very quick look, seems like it can't handle the process going away on its own: the actual view would be destroyed but the plugin's `procs[output].view` pointer would still point to its location at `fini()` time → use-after-free. @soreau 

(But why are you using that plugin with swaybg?? That plugin is for using a *regular* (xdg-shell) app as a background. swaybg already makes itself a background using layer-shell.)

**Comment by soreau** [comment]
@unrelentingtech Yes you are right. background-view needs a bit of attention, I want to change it to matching app-id instead of pid since app-id matching is more reliable.

**Comment by eternalinflation** [comment]
Thanks both for the update, for pointing me at xdg-shell, and apologies for the misleading refs, will do better next time ;)
When using e.g. mpv with `backgorund-view` instead it seems I can reproduce #1151.
